src/_includes/layouts/base.njk

```diff
@@ -30,6 +36,51 @@
             </article>
         </main>
         {% include "partials/global/footer.njk" %}
-        {% uioInit %}
+        {# {% uioInit %} #}
+        <script>
+        fluid.defaults("fluid.uiOptions.multilingual", {
+            gradeNames: ["fluid.uiOptions"],
+            distributeOptions: {
+                source: "{that}.options.locale",
+                target: "{that messageLoader}.options.resourceOptions.locale"
+            },
+            listeners: {
+                "onPrefsEditorReady.addLanguageAttributesToBody": {
+                    func: "fluid.uiOptions.multilingual.addLanguageAttributesToBody",
+                    args: ["{that}.prefsEditorLoader.prefsEditor.container", "{that}.options.locale", "{that}.options.direction"]
+                }
+            }
+        });
+        
+        /**
+        * Adds the locale and direction to the BODY in the IFRAME to enable CSS
+        * based on the locale and direction
+        *
+        * @param {jQuery} prefsEditorContainer - the DOM container for UIO
+        * @param {String} locale - the locale to apply
+        * @param {String} direction - the text orientation to apply
+        */
+        fluid.uiOptions.multilingual.addLanguageAttributesToBody = function (prefsEditorContainer, locale, direction) {
+            prefsEditorContainer.attr("lang", locale);
+            prefsEditorContainer.attr("dir", direction);
+        };
+
+        fluid.uiOptions.multilingual(".flc-prefsEditor-separatedPanel", {
+            auxiliarySchema: {
+                terms: {
+                    "templatePrefix": "/lib/infusion/src/framework/preferences/html",
+                    "messagePrefix": "/lib/infusion/src/framework/preferences/messages"
+                },
+                "fluid.prefs.tableOfContents": {
+                    enactor: {
+                        "tocTemplate": "/lib/infusion/src/components/tableOfContents/html/TableOfContents.html",
+                        "tocMessage": "/lib/infusion/src/framework/preferences/messages/tableOfContents-enactor.json"
+                    }
+                }
+            },
+            locale: "{{ lang | truncate(2, true, '') }}",
```

Comment:
Not sure I'm understanding this correctly. It looks like it will convert something like "fr-CA" to "fr", and pass that in as a locale. However, if there are multiple locales such as "fr-CA" and "fr-FR" it may not set the correct locale for UIO's message bundles. Perhaps you're doing this to reduce requests for UIO's template fetching for fallbacks?

Is there a way for the site builder (user of trivet) to add translations for UIO themselves? It's worth updating the README to mention how to do this. If it isn't possible, the README should probably mention that too, and also note which languages are available for UIO, maybe linking to the Infusion repo if appropriate.

Reply:
> Not sure I'm understanding this correctly. It looks like it will convert something like "fr-CA" to "fr", and pass that in as a locale. However, if there are multiple locales such as "fr-CA" and "fr-FR" it may not set the correct locale for UIO's message bundles. Perhaps you're doing this to reduce requests for UIO's template fetching for fallbacks?

Pretty much! As far as I can tell, UIO's only message bundles for country-specific languages are `en_US`, `en_CA` and `pt_BR`. In the interests of simplicity I opted to use the short (`en`) bundle for English which will allow Trivet and UIO to properly match up locales for all the other supported ones… but the Brazilian Portuguese locale makes this not as effective as I'd hoped it would be. 

> Is there a way for the site builder (user of trivet) to add translations for UIO themselves? It's worth updating the README to mention how to do this. If it isn't possible, the README should probably mention that too, and also note which languages are available for UIO, maybe linking to the Infusion repo if appropriate.

I think it would be much better for the Fluid ecosystem if site builders were encouraged to add translations to UIO directly so all other UIO integrators can benefit from them. Theoretically it would be possible for site builder to initialize UIO with a different messages path and copy all the translations they needed (and add some of their own) but I think the best bet is to indicate which locales are available in the README as per your suggestion.

Reply:
I've attempted to resolve this and eliminate the need for all that permalink determination logic by expanding the languages config block slightly in d2d694b. It now looks like this:

https://github.com/greatislander/trivet/blob/d2d694b97bdd19336f3f31d2b8e11b1f59d0a20a/README.md#L79-L88

Reply:
> Pretty much! As far as I can tell, UIO's only message bundles for country-specific languages are `en_US`, `en_CA` and `pt_BR`. In the interests of simplicity I opted to use the short (`en`) bundle for English which will allow Trivet and UIO to properly match up locales for all the other supported ones… but the Brazilian Portuguese locale makes this not as effective as I'd hoped it would be. I think at minimum I need to change my approach so that the replacement logic is to truncate OR change from kebab case to snake case depending on the locale in question. It's a bit messy but the problem with _not_ doing that is that in the example included in this repo, we have:
> 
> * en-CA
> * fr-CA
> 
> UIO includes:
> 
> * en_CA
> * fr
> 
> So I need to truncate one and replace the hyphen with an underscore in the other…

For starters we probably should have a fallback for Portugese if the region code isn't added. I've filed this as [FLUID-6675](https://issues.fluidproject.org/browse/FLUID-6675).

For "fr", if you pass in "fr_CA" it will fallback to "fr" so you don't actually have to truncate it. However, it would trigger additional requests to look for the "fr_CA" localization first. Considering where all this code runs, it would probably be better to truncate, or just add the "fr_CA" translation to UIO in Infusion. It probably already is "fr-CA" anyways, but @BlueSlug would be better to confirm that.

I can't remember why we use the "en_CA" vs "en-CA" format. Perhaps @amb26 would remember. Feel free to file a JIRA if you think that should be changed. It's something we could consider.

> I think it would be much better for the Fluid ecosystem if site builders were encouraged to add translations to UIO directly so all other UIO integrators can benefit from them. Theoretically it would be possible for site builder to initialize UIO with a different messages path and copy all the translations they needed (and add some of their own) but I think the best bet is to indicate which locales are available in the README as per your suggestion.

Yes, I agree adding more translations to UIO directly would be ideal. Although it's great to hear that it's possible for site builder to configure UIO directly if needed. I think we'll eventually want to add [crowdin](https://crowdin.com) or something like that for UIO, if possible. Although I'm going to guess this would have to wait till the Infusion 5.x timeline ( could be before the end of the year ). Until then, we'll likely be constrained to what's currently available.

Reply:
Is the `uioSlug` solution reasonable for now, do you think? It makes the mapping explicit which will save additional lookups/fallback logic.

Reply:
> For "fr", if you pass in "fr_CA" it will fallback to "fr" so you don't actually have to truncate it. However, it would trigger additional requests to look for the "fr_CA" localization first. Considering where all this code runs, it would probably be better to truncate, or just add the "fr_CA" translation to UIO in Infusion. It probably already is "fr-CA" anyways, but @BlueSlug would be better to confirm that.

IETF / BCP-47 Language Tags use a hyphen to separate the language part from the locale part, so it makes sense to be consistent with that specification. I can't recall why UIO/Infusion uses underscores, unfortunately.

Reply:
UIO went with underscores to be consistent with the Java locale format https://www.oracle.com/technical-resources/articles/javase/locale.html which was the common format for a lot of resource bundles at the time Infusion was being designed. We should switch to BCP-47 when we have the Infusion 5 rewrite.

Reply:
I've filed this as [FLUID-6676](https://issues.fluidproject.org/browse/FLUID-6676).